files/rsyslog/rsyslog.sh

```diff
@@ -228,6 +232,83 @@ fi
 # make sure there is not one left over from a previous run
 rm -f /var/run/rsyslogd.pid
 
+# Generate a config file to keep undefined fields
+keep_fields_lookup=${RSYSLOG_WORKDIRECTORY}/keep_fields.json
+keep_fields_conf=${RSYSLOG_WORKDIRECTORY}/10-keep_fields.conf
+rm -f $keep_fields_lookup $keep_fields_conf
+
+cat > $keep_fields_lookup << EOF
+{ "version" : 1,
+  "nomatch" : "unknown",
+  "type" : "string",
+  "table" : [
+EOF
+
+cat > $keep_fields_conf << EOF
+lookup_table(name="keep_fields" file="$keep_fields_lookup")
+
+template(name="debugTemplate" type="list") {
+    constant(value="\$!: ")
+    property(name="\$!")
+    constant(value="\\n")
+}
+
+template(name="firstFieldTemplate" type="list") {
+    constant(value="\\"") property(name="\$.ii!key") constant(value="\\":\\"")
+    property(name="\$.ii!value") constant(value="\\"")
+}
+
+template(name="addFieldTemplate" type="list") {
+    property(name="\$.undef")
+    constant(value=",\\"") property(name="\$.ii!key") constant(value="\\":\\"")
+    property(name="\$.ii!value") constant(value="\\"")
+}
+
+template(name="makeJsonTemplate" type="list") {
+    constant(value="{") property(name="\$.undef") constant(value="}")
+}
+
+# Moving undefined fields to \$!undefined
+if (strlen(\`echo \$CDM_USE_UNDEFINED\`) > 0) and (\`echo \$CDM_USE_UNDEFINED\` == "true") then {
+    set \$.field = "unknown";
+    set \$.firsttime = "true";
+    foreach (\$.ii in \$.all) do {
+        set \$.field = lookup("keep_fields", \$.ii!key);
+        if (strlen(\$.field) > 0) and (\$.field == "unknown") then {
+            if \$.firsttime == "true" then {
+                set \$.undef = exec_template('firstFieldTemplate');
+                set \$.firsttime = "false";
+            } else {
+                set \$.undef = exec_template('addFieldTemplate');
+            }
+        }
+    }
+    set \$!undefined = exec_template('makeJsonTemplate');
+}
+EOF
+
+for field in $( echo $DEFAULT_KEEP_FIELDS | sed -e "s/,/ /g" )
+do
+    cat >> $keep_fields_conf << EOF
+if strlen(\$.all!$field) > 0 then {
+    set \$!$field = \$.all!$field;
```

Comment:
Then after this line
`unset \$.all!$field;`
to "move" `$field` from `$.all` to `$!`